finagle-netty4/src/test/scala/com/twitter/finagle/netty4/Netty4SslTest.scala

```diff
@@ -106,33 +98,110 @@ class Netty4SslTest extends FunSuite with Eventually with IntegrationPatience {
       listener.listen(new InetSocketAddress(InetAddress.getLoopbackAddress, 0))(serveTransport(_))
     }
 
-    val clientConfig = SslClientConfiguration(
-      keyCredentials = KeyCredentials.CertAndKey(clientCert.certificate(), clientCert.privateKey()),
-      trustCredentials = TrustCredentials.CertCollection(serverCert.certificate())
-    )
-    val client = {
+    def server: ListeningServer
+
+      def mkClient(serverCert: X509Certificate, hostname: Option[String]): Service[String, String] = {
       val addr = server.boundAddress.asInstanceOf[InetSocketAddress]
-      new finagle.netty4.Netty4SslTest.Client(clientCert, serverCert).withTransport
-        .tls(clientConfig)
-        .newService(s"${addr.getHostName}:${addr.getPort}", "client")
+      new finagle.netty4.Netty4SslTest.Client().withTransport
+        .tls(SslClientConfiguration(hostname), new SslClientEngineFactory {
+        override def apply(address: Address, config: SslClientConfiguration): Engine = {
+          val ctx = SslContextBuilder.forClient.keyManager(clientCert.key(), clientCert.cert()).trustManager(serverCert).build()
+          new Engine(ctx.newEngine(allocator, hostname.get, -1))
+        }
+      }).newService(s"${addr.getHostName}:${addr.getPort}", "client")
     }
 
+    def client: Service[String, String]
+
     def close() = Future.collect(Seq(client.close(), server.close()))
   }
 
-  override type FixtureParam = Ctx
+  class DefaultCtx extends Ctx {
+    val serverCert = new SelfSignedCertificate("example.server.com")
+    val serverConfig = SslServerConfiguration(
+      keyCredentials = KeyCredentials.CertAndKey(serverCert.certificate(), serverCert.privateKey()),
+      trustCredentials = TrustCredentials.CertCollection(clientCert.certificate()),
+      clientAuth = ClientAuth.Needed)
+
+    override def server: ListeningServer = mkServer(Params.empty + ServerSsl(Some(serverConfig)))
+
+    override def client: Service[String, String] = mkClient(serverCert.cert(), Some("example.server.com"))
+  }
+
+  class SniCtx extends Ctx {
+
+    val serverCerts = Map(
+      "first.server.com" -> new SelfSignedCertificate("first.server.com"),
+      "second.server.com" -> new SelfSignedCertificate("second.server.com")
+    )
+
+    private def mkConfig(serverCert: SelfSignedCertificate): SslServerConfiguration = {
+      SslServerConfiguration(
+        keyCredentials = KeyCredentials.KeyAndCertChain(serverCert.key(), Array(serverCert.cert())),
+        trustCredentials = TrustCredentials.Certificates(Array(clientCert.cert())),
+        clientAuth = ClientAuth.Needed
+      )
+    }
+
+    override def server: ListeningServer =
+      mkServer(Params.empty + ServerSsl(Some(SslServerConfiguration(clientAuth = ClientAuth.Needed))) +
+        SniSupport.fromOption(serverCerts.get(_).map(mkConfig)))
+
+    override def client: Service[String, String] = {
+      val (hostname, serverCrt) = serverCerts.head
+      mkClient(serverCrt.cert(), Some(hostname))
```

Comment:
Change mkClient to just take hostname: String instead of hostname: Option[String] here, so that it's not necessary to calling hostname.get later.